File: CS5030/Notes/practice.py
def balance_interations(n, p):
    n= 15 # number of iterations
    p= 3 # number of cores
    # Total time on each core
    time = [0]*p # This will create a list of p zeros.

    print ("Initial time:", time)

    iterations = [] # compile a list of iterations for each core. (list of lists)
    for i in range(p):
        iterations.append([])
    print ("Initial iterations:", iterations)

    # Now that we have our data structures, we can start assigning iterations to cores.

    for i in range(n, 0, -1): # start from n downt o 1
        # find core with minimum time
        min_core = time.index(min(time))

        # assign iteration i to that core
        iterations[min_core].append(i)
        time[min_core] += i # update total time for that core (add i)
        print("Final time:", time)
        print("Iterations assigned to each core:", iterations)


balance_interations(15, 3)

File: CS5030/Notes/practice.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>

int main(){
    int n = 15;
    int p = 3;

    // Total time on each core
    std::vector<int> time(p,0); // this creates a vector of size p with all 0s named time

    // COmpile a list of iterations for each core (list of lists)
    std::vector<std::vector<int>>iterations(p); // this creates a vector of size p with empty vectors inside named iterations 

    // Now that we have out data structures we can start assigning iterations to cores

    for(int i = n; i>=0; i--){
        // find core with minimum time
        int min_core = std::min_element(time.begin(), time.end()) - time.begin();

        // assign iteration i to min_core
        iterations[min_core].push_back(i);

        // update time for min_core
        time[min_core] += i;


    }

    // Print out the assignments
    for (int core = 0; core<p; core++){ // loop starting at core =0 while core is less than p
        std::cout << "Core " << core << ": (time = " << time[core] << ") " ;
        for (int it : iterations[core]){ // range based for loop
            std::cout << it << " ";
        
        }
        std::cout << std::endl;
    }
    

}


